smart_open/gcs.py

```diff
@@ -0,0 +1,523 @@
+            new_position = self._current_pos + offset
+        else:
+            new_position = self._size + offset
+        new_position = clamp(new_position, 0, self._size)
+        self._current_pos = new_position
+        self._raw_reader.seek(new_position)
+        logger.debug('new_position: %r', self._current_pos)
+
+        self._buffer.empty()
+        self._eof = self._current_pos == self._size
+        return self._current_pos
+
+    def tell(self):
+        """Return the current position within the file."""
+        return self._current_pos
+
+    def truncate(self, size=None):
+        """Unsupported."""
+        raise io.UnsupportedOperation
+
+    def read(self, size=-1):
+        """Read up to size bytes from the object and return them."""
+        if size == 0:
+            return b''
+        elif size < 0:
+            from_buf = self._read_from_buffer()
+            self._current_pos = self._size
+            return from_buf + self._raw_reader.read()
+
+        #
+        # Return unused data first
+        #
+        if len(self._buffer) >= size:
+            return self._read_from_buffer(size)
+
+        #
+        # If the stream is finished, return what we have.
+        #
+        if self._eof:
+            return self._read_from_buffer()
+
+        #
+        # Fill our buffer to the required size.
+        #
+        self._fill_buffer(size)
+        return self._read_from_buffer(size)
+
+    def read1(self, size=-1):
+        """This is the same as read()."""
+        return self.read(size=size)
+
+    def readinto(self, b):
+        """Read up to len(b) bytes into b, and return the number of bytes
+        read."""
+        data = self.read(len(b))
+        if not data:
+            return 0
+        b[:len(data)] = data
+        return len(data)
+
+    def readline(self, limit=-1):
+        """Read up to and including the next newline.  Returns the bytes read."""
+        if limit != -1:
+            raise NotImplementedError('limits other than -1 not implemented yet')
+        the_line = io.BytesIO()
+        while not (self._eof and len(self._buffer) == 0):
+            #
+            # In the worst case, we're reading the unread part of self._buffer
+            # twice here, once in the if condition and once when calling index.
+            #
+            # This is sub-optimal, but better than the alternative: wrapping
+            # .index in a try..except, because that is slower.
+            #
+            remaining_buffer = self._buffer.peek()
+            if self._line_terminator in remaining_buffer:
+                next_newline = remaining_buffer.index(self._line_terminator)
+                the_line.write(self._read_from_buffer(next_newline + 1))
+                break
+            else:
+                the_line.write(self._read_from_buffer())
+                self._fill_buffer()
+        return the_line.getvalue()
+
+    def terminate(self):
+        """Do nothing."""
+        pass
```

Comment:
Is this OK?

Theoretically, you can continue reading after calling terminate, which doesn't sound very useful. Why is this function even needed?